sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala

```diff
@@ -69,11 +70,28 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
     checkAnswer(
       df.select("t").filter($"t" <= "2014-06-01"),
       Row(Timestamp.valueOf("2014-01-01 00:00:00")) :: Nil)
+  }
 
+  test("timestamp comparison with timestamp strings with session local timezone") {
+    val df = Seq(
+      (1, Timestamp.valueOf("2015-12-31 16:00:00")),
+      (2, Timestamp.valueOf("2016-01-01 00:00:00"))).toDF("i", "t")
 
     checkAnswer(
-      df.select("t").filter($"t" >= "2014-06-01"),
-      Row(Timestamp.valueOf("2015-01-01 00:00:00")) :: Nil)
+      df.select("t").filter($"t" <= "2016-01-01 00:00:00"),
+      Seq(
+        Row(Timestamp.valueOf("2015-12-31 16:00:00")),
+        Row(Timestamp.valueOf("2016-01-01 00:00:00"))))
+
+    withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> "GMT") {
+
+      // $"t" string in GMT would be as follows respectively:
+      // "2016-01-01 00:00:00"
+      // "2016-01-01 08:00:00"
+      checkAnswer(
+        df.select("t").filter($"t" <= "2016-01-01 00:00:00"),
+        Row(Timestamp.valueOf("2015-12-31 16:00:00")))
```

Comment:
this shows that it will be very confusing if the session local timezone is different from JVM default timezone in driver...